riscv/src/code_gen.rs

```diff
@@ -319,43 +375,154 @@ fn preamble<T: FieldElement>(runtime: &Runtime, with_bootloader: bool) -> String
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - val2_col + Z,
+        pc' = pc + 1 + (XXIsZero * W)
+    }
 
     // input X is required to be the difference of two 32-bit unsigend values.
     // i.e. -2**32 < X < 2**32
-    instr branch_if_positive X, l: label {
-        X + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
+    instr branch_if_positive X, Y, Z, l: label
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        (val1_col - val2_col + Z) + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
         pc' = wrap_bit * l + (1 - wrap_bit) * (pc + 1)
     }
     // input X is required to be the difference of two 32-bit unsigend values.
     // i.e. -2**32 < X < 2**32
-    instr is_positive X -> Y {
-        X + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
-        Y = wrap_bit
+    // reg1 - reg2 + offset
+    instr is_positive X, Y, Z, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+        link ~> regs.mstore(W, STEP + 2, val3_col)
+    {
+        (val1_col - val2_col + Z) + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
+        val3_col = wrap_bit
+    }
+
+    instr move_reg X, Y, Z, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> regs.mstore(Y, STEP + 1, val3_col)
+    {
+        val3_col = val1_col * Z + W
+    }
+
+    // ======aaaaaaaaaaaa=========
+
+    // Wraps a value in Y to 32 bits.
+    // Requires 0 <= Y < 2**33
+    // These are the old `wrap` instruction.
+    
+    instr add_new X, Y, Z, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+        link ~> regs.mstore(W, STEP + 2, val3_col)
+    {
+        val1_col + val2_col + Z = val3_col + wrap_bit * 2**32, val3_col = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000
+    }
+
+    // Requires -2**32 <= Y < 2**32
+    // These are the old `wrap_signed` instruction.
+    instr add_new_signed X, Y, Z, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+        link ~> regs.mstore(W, STEP + 2, val3_col)
+    {
+        (val1_col - val2_col + Z) + 2**32 = val3_col + wrap_bit * 2**32, val3_col = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000
     }
 
     // ================= logical instructions =================
 
-    instr is_equal_zero X -> Y { Y = XIsZero }
-    instr is_not_equal_zero X -> Y { Y = 1 - XIsZero }
+    instr is_equal_zero X, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> regs.mstore(W, STEP + 2, val3_col)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col,
+        val3_col = XXIsZero
```

Comment:
```suggestion
        link ~> XX = regs.mload(X, STEP)
        link ~> regs.mstore(W, STEP + 2, XXIsZero)
    {
        XXIsZero = 1 - XX * XX_inv,
```
?

Reply:
I find it less explicit and readable, but fine